File: src/client/zig/dom.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

const env = @import("browser_env.zig");
const callback = @import("callback.zig");
const browser = @import("browser.zig");
const enums = @import("browser_enums.zig");
const InsertPosition = enums.InsertPosition;
const EventType = enums.EventType;

pub fn getElementById(id: []const u8) i32 {
    return env.getElementById(id.ptr, id.len);
}

pub fn setShown(handle: i32, shown: bool) void {
    env.setElementShown(handle, @boolToInt(shown));
}

pub fn setTextContent(handle: i32, text: []const u8) void {
    env.setElementTextContent(handle, text.ptr, text.len);
}

pub const getChildrenCount = env.getChildrenCount;
pub const getChild = env.getChild;

pub fn insertAdjacentHTML(handle: i32, position: InsertPosition, html: []const u8) void {
    env.insertAdjacentHTML(handle, position, html.ptr, html.len);
}
pub const removeLastChild = env.removeLastChild;

pub fn addClass(handle: i32, class: []const u8) void {
    env.addClass(handle, class.ptr, class.len);
}
pub fn removeClass(handle: i32, class: []const u8) void {
    env.removeClass(handle, class.ptr, class.len);
}
pub fn setAttribute(handle: i32, key: []const u8, value: []const u8) void {
    env.setAttribute(handle, key.ptr, key.len, value.ptr, value.len);
}
pub fn getAttribute(handle: i32, allocator: *Allocator, key: []const u8) []u8 {
    const packed_slice = env.getAttribute(
        handle,
        callback.allocator(allocator).handle,
        key.ptr,
        key.len,
    );
    return browser.unpackSlice(packed_slice);
}

pub fn searchAncestorsForClass(start_handle: i32, stop_handle: i32, class: []const u8) i32 {
    return env.searchAncestorsForClass(start_handle, stop_handle, class.ptr, class.len);
}

pub fn addEventListener(
    handle: i32,
    event_type: EventType,
    cb: callback.CallbackI32,
) void {
    return env.addEventListener(handle, event_type, cb.handle);
}
pub fn addWindowEventListener(
    event_type: EventType,
    cb: callback.CallbackI32,
) void {
    return env.addWindowEventListener(event_type, cb.handle);
}
pub const getEventTarget = env.getEventTarget;
pub const getEventModifiers = env.getEventModifiers;
pub const getKeyboardEventCode = env.getKeyboardEventCode;
pub const preventDefault = env.preventDefault;
pub const stopPropagation = env.stopPropagation;

pub fn setInputValue(handle: i32, value: []const u8) void {
    return env.setInputValue(handle, value.ptr, value.len);
}
pub fn getInputValue(handle: i32, allocator: *Allocator) []u8 {
    const packed_slice = env.getInputValue(
        handle,
        callback.allocator(allocator).handle,
    );
    return browser.unpackSlice(packed_slice);
}
pub const setInputValueAsNumber = env.setInputValueAsNumber;
pub const getInputValueAsNumber = env.getInputValueAsNumber;

pub const focus = env.focus;
pub const blur = env.blur;

File: src/client/js/dom.js

const {HandleRegistry} = require("handleRegistry");
const {KeyboardEventCode_inverse} = require("enums");

const elementRegistry = new HandleRegistry();
const eventRegistry = new HandleRegistry();

function getElementByHandle(handle) {
    let element = elementRegistry.registry[handle];
    if (element != null) {
        return element;
    }
    element = document.getElementById("h-" + handle.toString(16));
    if (element != null) {
        return element;
    }
    throw new Error("bogus element handle: 0x" + handle.toString(16));
}

function getElementHandle(element) {
    let handle_str = element.getAttribute("data-h");
    if (handle_str != null) {
        return parseInt(handle_str, 16);
    }
    if (element.getAttribute("id") != null || !document.contains(element)) {
        // using document.getElementById("handle-" + handle) isn't going to work.
        // Use the registry to find this element later.
        const {handle} = elementRegistry.alloc(element);
        handle_str = handle.toString(16);
        element.setAttribute("data-h", handle_str);
        return handle;
    } else {
        // Give the element an id to use to find it later.
        // This avoids pointers to deleted elements preventing them from being garbaged collected.
        const handle = elementRegistry.nextHandle();
        handle_str = handle.toString(16);
        element.setAttribute("id", "h-" + handle_str);
        element.setAttribute("data-h", handle_str);
        return handle;
    }
}

function getElementById(id) {
    const element = document.getElementById(id);
    if (element == null) throw new Error("no such element: " + id);
    return getElementHandle(element);
}

function setElementShown(handle, shown) {
    getElementByHandle(handle).style.display = shown ? "" : "none";
}

function setElementTextContent(handle, text) {
    getElementByHandle(handle).textContent = text;
}

function getChildrenCount(handle) {
    return getElementByHandle(handle).children.length;
}

function getChild(parent_handle, i) {
    const parent = getElementByHandle(parent_handle);
    const child = parent.children[i];
    if (child == null) throw new Error("bad child index: " + i);
    return getElementHandle(child);
}

function insertAdjacentHTML(handle, position, html) {
    getElementByHandle(handle).insertAdjacentHTML(position, html);
}

function removeLastChild(handle) {
    const element = getElementByHandle(handle);
    element.removeChild(element.lastChild);
}

function addClass(handle, class_) {
    getElementByHandle(handle).classList.add(class_);
}
function removeClass(handle, class_) {
    getElementByHandle(handle).classList.remove(class_);
}
function setAttribute(handle, key, value) {
    getElementByHandle(handle).setAttribute(key, value);
}
function getAttribute(handle, key) {
    return getElementByHandle(handle).getAttribute(key);
}

// Searches the start_handle node and its .parentNode ancestors for a node whose .classList.contains(class_).
// If not found, returns stop_handle.
function searchAncestorsForClass(start_handle, stop_handle, class_) {
    let node = getElementByHandle(start_handle);
    if (node.classList.contains(class_)) return start_handle;
    const root = getElementByHandle(stop_handle);
    if (!root.contains(node)) return stop_handle;
    if (node === root) return stop_handle;

    while (true) {
        node = node.parentNode;
        if (node === root) return stop_handle;
        if (node.classList.contains(class_)) {
            return getElementHandle(node);
        }
    }
}

function addEventListener(handle, event_type, cb) {
    getElementByHandle(handle).addEventListener(event_type, function(event) {
        const {handle, dispose} = eventRegistry.alloc(event);
        try {
            cb(handle);
        } finally {
            dispose();
        }
    });
}

function addWindowEventListener(event_type, cb) {
    window.addEventListener(event_type, function(event) {
        const {handle, dispose} = eventRegistry.alloc(event);
        try {
            cb(handle);
        } finally {
            dispose();
        }
    });
}

function getEventTarget(event_handle) {
    const event = eventRegistry.registry[event_handle];
    const target = event.target;
    return getElementHandle(target);
}
function getEventModifiers(event_handle) {
    const event = eventRegistry.registry[event_handle];
    return (
        (event.shiftKey ? (1 << 0) : 0) |
        (event.ctrlKey ? (1 << 1) : 0) |
        (event.altKey ? (1 << 2) : 0) |
        (event.metaKey ? (1 << 3) : 0)
    );
}
function getKeyboardEventCode(event_handle) {
    const event = eventRegistry.registry[event_handle];
    const code_code = KeyboardEventCode_inverse[event.code];
    if (typeof code_code !== "number") throw new Error("event.code not recognized: " + event.code);
    return code_code;
}
function preventDefault(event_handle) {
    const event = eventRegistry.registry[event_handle];
    event.preventDefault();
}

function setInputValueAsNumber(handle, value) {
    getElementByHandle(handle).valueAsNumber = value;
}
function getInputValueAsNumber(handle) {
    return getElementByHandle(handle).valueAsNumber;
}

return {
    getElementByHandle,
    getElementHandle,
    getElementById,
    setElementShown,
    setElementTextContent,
    getChildrenCount,
    getChild,
    insertAdjacentHTML,
    removeLastChild,
    addClass,
    removeClass,
    setAttribute,
    getAttribute,
    searchAncestorsForClass,
    addEventListener,
    addWindowEventListener,
    getEventTarget,
    getEventModifiers,
    getKeyboardEventCode,
    preventDefault,
    setInputValueAsNumber,
    getInputValueAsNumber,
};
